src/LEGO.AsyncAPI.Readers/AsyncApiExternalReferenceResolver.cs

```diff
@@ -0,0 +1,239 @@
+        {
+            this.ResolveObject(schema.Items, r => schema.Items = r);
+            this.ResolveList(schema.OneOf);
+            this.ResolveList(schema.AllOf);
+            this.ResolveList(schema.AnyOf);
+            this.ResolveObject(schema.Contains, r => schema.Contains = r);
+            this.ResolveObject(schema.Else, r => schema.Else = r);
+            this.ResolveObject(schema.If, r => schema.If = r);
+            this.ResolveObject(schema.Items, r => schema.Items = r);
+            this.ResolveObject(schema.Not, r => schema.Not = r);
+            this.ResolveObject(schema.Then, r => schema.Then = r);
+            this.ResolveObject(schema.PropertyNames, r => schema.PropertyNames = r);
+            this.ResolveObject(schema.AdditionalProperties, r => schema.AdditionalProperties = r);
+            this.ResolveMap(schema.Properties);
+        }
+
+        private void ResolveObject<T>(T entity, Action<T> assign)
+            where T : class, IAsyncApiReferenceable, new()
+        {
+            if (entity == null)
+            {
+                return;
+            }
+
+            if (this.IsUnresolvedReference(entity))
+            {
+                assign(this.ResolveReference<T>(entity.Reference));
+            }
+        }
+
+        private void ResolveList<T>(IList<T> list)
+            where T : class, IAsyncApiReferenceable, new()
+        {
+            if (list == null)
+            {
+                return;
+            }
+
+            for (int i = 0; i < list.Count; i++)
+            {
+                var entity = list[i];
+                if (this.IsUnresolvedReference(entity))
+                {
+                    list[i] = this.ResolveReference<T>(entity.Reference);
+                }
+            }
+        }
+
+        private void ResolveMap<T>(IDictionary<string, T> map)
+            where T : class, IAsyncApiReferenceable, new()
+        {
+            if (map == null)
+            {
+                return;
+            }
+
+            foreach (var key in map.Keys.ToList())
+            {
+                var entity = map[key];
+                if (this.IsUnresolvedReference(entity))
+                {
+                    map[key] = this.ResolveReference<T>(entity.Reference);
+                }
+            }
+        }
+
+        private T ResolveReference<T>(AsyncApiReference reference)
+            where T : class, IAsyncApiReferenceable, new()
+        {
+            if (reference.IsExternal)
+            {
+                if (this.readerSettings.ExternalReferenceReader is null)
+                {
+                    throw new AsyncApiReaderException(
+                        "External reference configured in AsyncApi document but no implementation provided for ExternalReferenceReader.");
+                }
+
+                // read external content
+                var externalContent = this.readerSettings.ExternalReferenceReader.GetExternalResource(reference.Reference);
```

Comment:
Im guessing this means that references to fragments must be handled by the referenceReader itself.
At some point, it would be interesting to handle this ourselves, without pushing it to the implementor.
So all the ReferenceReader needs to do, is to get the referenced document.

Reply:
I agree. I'm wondering if may we should collab on a roadmap or a capability matrix or something, just so we're tracking these sorts of feature ideas?

Reply:
Resolving for now as it sounds like more follow up work.